File: arrays/static/StaticArray.java
/**
* Basic implementation of a static array in Java
* 
* :author: Pierre Bouillon [https://pbouillon.github.io/]
* :licence: MIT [https://github.com/pBouillon/data_structures/blob/master/LICENSE]
*/

public class StaticArray<T> {

    /**
     * 
     */
    private int _nbElements;
    
    /**
     * 
     */
    private int _size;
    
    /**
     * 
     */
    private T[] _values;

    /**
     * 
     * @param capacity
     */
    public StaticArray (int capacity) {
        if (capacity < 0) {
            throw new IllegalArgumentException("Array size must be a positive number");
        }

        _nbElements = 0;
        _size = capacity;
        _values = (T[]) new Object[_size];
    } /* */

    /**
     * 
     */
    public void clear () {
        _nbElements = 0;
        for (int i = 0; i < _size; ++i) {
            _values[i] = null;
        }
    } /* */

    /**
     * 
     * @param value
     * @return
     */
    public boolean contains (T value) {
        for (int i = 0; i < _size; ++i) {
            if (_values[i] == value) {
                return true;
            }
        }
        return false ;
    } /* */

    /**
     * 
     * @param index
     * @return
     */
    public T get (int index) {
        if (index < 0 || index > _size - 1) {
            throw new IllegalArgumentException("Index out of range");
        }

        return _values[index];
    } /* */

    /**
     * 
     * @param value
     */
    public void initialize (T value) {
        _nbElements = _size;

        for (int i = 0; i < _size; ++i) {
            _values[i] = value;
        }
    } /* */

    /**
     * 
     * @return
     */
    public boolean isEmpty () {
        return _nbElements == 0;
    } /* */

    /**
     * 
     * @return
     */
    public int size() {
        return _size;
    } /* */

    /**
     * 
     * @param index
     * @param value
     */
    public void set (int index, T value) {
        if (index < 0 || index > _size - 1) {
            throw new IllegalArgumentException("Index out of range");
        }

        if (value != null) {
            ++_nbElements;
        }
        else if (value == null && _values[index] != null) {
            --_nbElements;
        }

        _values[index] = value;
    } /* */
}

File: arrays/static/StaticArray.cs
/**
 * Basic implementation of a static array in C#
 * 
 * @author: Pierre Bouillon [https://pbouillon.github.io/]
 * @licence: MIT [https://github.com/pBouillon/data_structures/blob/master/LICENSE]
 */

using System;

namespace DataStructures
{
    /// <summary>
    /// Implements a basic static array
    /// </summary>
    /// <typeparam name="T">Type of the desired array</typeparam>
    class StaticArray<T>
    {
        private T[] _values;
        private readonly int _capacity;

        /// <summary>
        /// Default constructor
        /// </summary>
        /// <param name="capacity">Max array capacity</param>
        /// <param name="initialValue">Value with whom the array will be initialized</param>
        public StaticArray(int capacity, T initialValue = default)
        {
            _capacity = capacity;
            _values = new T[_capacity];

            for (var i = 0; i < _capacity; ++i)
            {
                _values[i] = initialValue;
            }
        }

        /// <summary>
        /// Default accessors for an array
        /// </summary>
        /// <param name="index">index of the cell to access</param>
        /// <returns>The value on the given index</returns>
        public T this[int index]
        {
            get => _values[index];
            set => _values[index] = value;
        }

        /// <summary>
        /// Add a value at a specific index
        /// </summary>
        /// <param name="value">Value to add</param>
        /// <param name="index">Index of the cell to update</param>
        public void Add(T value, int index)
        {
            if (index < 0 || index >= _capacity)
            {
                throw new ArgumentOutOfRangeException();
            }

            _values[index] = value;
        }

        /// <summary>
        /// Reset all values of the array
        /// </summary>
        public void Clear()
        {
            _values = new T[_capacity];
        }

        /// <summary>
        /// Get an element at a specific index
        /// </summary>
        /// <param name="index">Index of the cell we are looking for</param>
        /// <returns>The element in the designated cell</returns>
        public T Get(int index)
        {
            if (index < 0 || index >= _capacity)
            {
                throw new ArgumentOutOfRangeException();
            }

            return _values[index];
        }
    }
}
